### PR TITLE
KAFKA-9487: Follow-up PR of Kafka-9445

### PR DESCRIPTION
**Pull request by brary**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by guozhangwang — COMMENTED** [review]
The non unit-testing part LGTM. @mjsax could you take a look at the unit tests and see if they have addressed all your comments?

File: streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java
Comment (vinothchandar) [inline streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java]: cosmetic: `Represents all the query options that a user can provide to control the kind of stores we would like to fetch` ?

File: streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java
Comment (vinothchandar) [inline streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java]: Cosmetic: ` Options could indicate - whether we want stale stores to be included (or) we want to filter for only stores belonging to a partition`.

File: streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java
Comment (vinothchandar) [inline streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java]: Could we use a different sentinel than `null` for the `partition` field?. Does an `Optional` make sense?

File: streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java
Comment (vinothchandar) [inline streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java]: nit: we need the `*` before `@return` right?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (vinothchandar) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: any reason we can't use an empty list for the sentinel?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (vinothchandar) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: can't we just lookup the map? why iterate over if we are going to compare the key anyway?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (vinothchandar) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: a side note: Should we also consider the map `stateStoreNameToSourceRegex` here?

**Review by vinothchandar — COMMENTED** [review]
Left few comments..

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (brary) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: I think the return type of stateStoreNameToSourceTopics is a set and we wanted to return a list, so I had to loop and add it to list.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (brary) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: Done.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (brary) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: I am not sure, do you think it will identify more topic groups? At the end of the day we are identifying a topic groupId to which a store belongs to. 

On a side note, if we do want to add stateStoreNameToSourceRegex along with stateStoreNameToSourceTopics, then we can probably return a set instead of list in this function and iterate over both of above one by one.

File: streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java
Comment (brary) [reply streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java]: Done.

File: streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java
Comment (brary) [reply streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java]: Done.

File: streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java
Comment (brary) [reply streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java]: Done.

File: streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java
Comment (brary) [reply streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java]: Actually we had a long discussion on keeping this null and adding it in Java doc that null means all partitions. Do you think this is important to start a discussion on the thread again? :)

**Comment by brary** [comment]
> Left few comments..

Thanks @guozhangwang and @vinothchandar . Addressed a few comments and have replied to some left.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]:
```suggestion
     * @param storeQueryParams   the parameters used to fetch a queryable store
```

File: streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java]: This comment can easily become outdated as the class evolved. Better to keep it high level and defer to the setters' javadocs for a description of the available operations.

Eg:
`{@code StoreQueryParams} allows you to pass a variety of parameters when fetching a store for interactive query.`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: I'm also a little confused. Would this not be the same?
```
results.addAll(maybeDecorateInternalSourceTopics(stateStoreNameToSourceTopics.get(storeName))
```
I.e., converting the returned set to a list doesn't seem to require iterating over the whole map just to look for one key.

On a different note, it doesn't seem like we really need a List in particular. We could just relax it to Collection, and then you wouldn't have to convert the set to a list. But that's not a big issue to me.

**Review by vvcephei — CHANGES_REQUESTED** [review]
Hey, thanks for the update @brary .

IIUC, the main motivation for this PR was some follow-up comments from @mjsax . It would be good to get a +1 from him before merging. Maybe after the other comments are addressed.

File: streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java
Comment (brary) [reply streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java]: Done.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (brary) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Added this my commit for addressing the other comments.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (brary) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: Makes sense, Changed to Collection and removed the loop.

**Comment by brary** [comment]
> Hey, thanks for the update @brary .
> 
> IIUC, the main motivation for this PR was some follow-up comments from @mjsax . It would be good to get a +1 from him before merging. Maybe after the other comments are addressed.

Thanks @vvcephei for the comments. I have made the changes suggested by you as well. 

@mjsax  I guess everyone else has reviewed and their comments have been incorporated. It would be great if you can give one last look and if it's okay merge it.

File: streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java
Comment (vinothchandar) [reply streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java]: dont worry about it :) .. Happy to budge on stylistic things and leave it to the committers' call

File: streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java]: Nit: No need to do method calls, we can just access the variables directly.  Also, try to avoid unnecessary `this.` prefix

File: streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java]: as above.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProvider.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProvider.java]: I cannot comment below, but I think we should still simplify the code when the user specifies a partition. I left a comment on the original PR about it (https://github.com/apache/kafka/pull/7984/files#r375513526). Let me know if this comment makes sense.

**Review by mjsax — COMMENTED** [review]
Thanks for the PR. Couple of minor comments. Also added one more reply on the original PR that should be addressed in this PR (https://github.com/apache/kafka/pull/7984/files#r375517935).

**Review by vvcephei — COMMENTED** [review]
Thanks for addressing my comments!

**Comment by mjsax** [comment]
@vinothchandar @vvcephei I would still propose to rename the new class to `StoreQueryParameters` -- it's not a big deal to rename even after feature freeze IHMO.

Thoughts?

File: streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java
Comment (brary) [reply streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java]: Done.

File: streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java
Comment (brary) [reply streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java]: Done.

**Comment by vinothchandar** [comment]
`StoreQueryParameters` does sound better to me. but leave it you all. :)

File: streams/src/main/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProvider.java
Comment (brary) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProvider.java]: Thanks. It makes sense, I have made the suggested change.

**Comment by brary** [comment]
@mjsax  I have changed the class to StoreQueryParameters and addressed the comment you gave in earlier PR. There were few checkstyles which were failing in ValueAndTimestampSerializerTest unrelated to this PR. I have fixed those as well.

**Review by mjsax — APPROVED** [review]
Thanks @brary! LGTM. I leave it to @vvcephei to merge this PR to `trunk` and cherry-pick to `2.5` branch.

**Comment by vvcephei** [comment]
Thanks @mjsax . I'll take care of it tomorrow.

I don't see any test results, so I'll ask Jenkins to re-run it:

Retest this, please.

**Comment by vvcephei** [comment]
Unrelated test failures:
kafka.admin.DeleteConsumerGroupsTest.testDeleteCmdAllGroups
kafka.admin.DeleteConsumerGroupsTest.testDeleteCmdWithMixOfSuccessAndError

Proceeding to merge.

**Comment by vvcephei** [comment]
Cherry-picked onto 2.5 as 3e77458cfb0f7a7586f17777f0e775d91e4744e1

**Pull request merged**
